File: LeetCode/cpp/599_minimum_index_sum_of_two_lists.cc
/*
Difficulty: Easy

Runtime: 79 ms (95%)
Memory: 35.6 MB (75%)
*/

#include <algorithm>
#include <cstddef>
#include <limits>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

class Solution {
 public:
  /// Time: O(n + m), Space: O(min(n, m))
  std::vector<std::string> findRestaurant(std::vector<std::string>& list1,
                                          std::vector<std::string>& list2) {
    std::vector<std::string>& shorter =
        list1.size() < list2.size() ? list1 : list2;
    std::vector<std::string>& longer =
        list1.size() < list2.size() ? list2 : list1;

    std::unordered_map<std::string, std::size_t> restaurantIndices;
    for (std::size_t i = 0; i < shorter.size(); ++i) {
      restaurantIndices.emplace(std::move(shorter[i]), i);
    }

    std::unordered_map<std::string, std::size_t> commonRestaurantSums;
    std::size_t minSum = std::numeric_limits<std::size_t>::max();

    for (std::size_t i = 0; i < longer.size(); ++i) {
      std::string& currRestaurant = longer[i];
      const auto it = restaurantIndices.find(currRestaurant);

      if (it != restaurantIndices.cend()) {
        const std::size_t sumIdx = i + it->second;
        minSum = std::min(minSum, sumIdx);
        commonRestaurantSums.emplace(std::move(currRestaurant), sumIdx);
      }
    }

    std::vector<std::string> bestCommonFavorites;
    for (const auto& [restaurant, sum] : commonRestaurantSums) {
      if (sum == minSum) {
        bestCommonFavorites.push_back(restaurant);
      }
    }

    return bestCommonFavorites;
  }
};

File: LeetCode/rust/599_minimum_index_sum_of_two_lists.rs
/*
Difficulty: Easy
*/

use std::cmp;
use std::collections::HashMap;

impl Solution {
    pub fn find_restaurant(list1: Vec<String>, list2: Vec<String>) -> Vec<String> {
        let (shorter, longer) = match list1.len() < list2.len() {
            true => (list1, list2),
            false => (list2, list1),
        };
        let restaurant_indices: HashMap<String, usize> = shorter
            .into_iter()
            .enumerate()
            .map(|(i, restaurant)| (restaurant, i))
            .collect();

        let mut min_sum = usize::MAX;
        let mut common_restaurant_sums = HashMap::<String, usize>::new();

        for (i, restaurant) in longer.into_iter().enumerate() {
            if let Some(&pair_idx) = restaurant_indices.get(&restaurant) {
                let idx_sum = i + pair_idx;
                min_sum = cmp::min(min_sum, idx_sum);
                common_restaurant_sums.insert(restaurant, idx_sum);
            }
        }

        common_restaurant_sums
            .into_iter()
            .filter(|(_, sum)| *sum == min_sum)
            .map(|(restaurant, _)| restaurant)
            .collect()
    }
}
